src/toxcore/dht/server/hole_punching.rs

```diff
@@ -0,0 +1,433 @@
+    /// last timestamp of trying hole punch
+    /// it is used to send NatPingRequest every 3 seconds
+    pub last_punching_time: Instant,
+    /// factor variable for guessing NAT port
+    pub first_punching_index: u32,
+    /// another factor variable for guessing NAT port
+    pub last_punching_index: u32,
+    /// holds ping_id used in NatPingRequest
+    /// multi NatPingRequest has this same ping_id
+    /// because every NatPingRequest receives NatPingResponse.
+    pub ping_id: u64,
+}
+
+impl HolePunching {
+    /// new object of HolePunching
+    pub fn new() -> Self {
+        HolePunching {
+            is_punching_done: true,
+            num_punch_tries: 0,
+            last_recv_ping_time: Instant::now(),
+            last_send_ping_time: Instant::now(),
+            last_punching_time: Instant::now(),
+            first_punching_index: 0,
+            last_punching_index: 0,
+            ping_id: HolePunching::new_ping_id(),
+        }
+    }
+
+    // get new ping id for NatPingRequest packet
+    fn new_ping_id() -> u64 {
+        loop {
+            let ping_id = random_u64();
+            if ping_id != 0 {
+                return ping_id;
+            }
+        }
+    }
+
+    /// send NatPingRequest and if condition is true, do hole punch
+    pub fn try_nat_punch(&mut self, server: &Server, friend_pk: PublicKey, addrs: Vec<SocketAddr>,
+                         nat_ping_req_interval: Duration) -> IoFuture<()> {
+        if !self.is_punching_done &&
+            self.last_punching_time.elapsed() >= nat_ping_req_interval &&
+            self.last_recv_ping_time.elapsed() <= nat_ping_req_interval * 2 {
+                let ip = match HolePunching::get_major_ip(&addrs, MAX_CLIENTS_PER_FRIEND / 2) {
+                    None => return Box::new(future::ok(())),
+                    Some(ip) => ip,
+                };
+
+                if self.last_punching_time.elapsed() > Duration::from_secs(RESET_PUNCH_INTERVAL) {
+                    self.num_punch_tries = 0;
+                    self.first_punching_index = 0;
+                    self.last_punching_index = 0;
+                }
+
+                let ports_to_try = HolePunching::get_nat_ports(&addrs, ip);
+
+                let res = self.punch(ports_to_try, ip, server, friend_pk);
+
+                self.last_punching_time = Instant::now();
+                self.is_punching_done = true;
+
+                res
+        } else {
+            // NatPingResponse is not responded, or hole punching was already done,
+            // or the elapsed time is too long since we received NatPingResponse,
+            // then we do nothing.
+            Box::new(future::ok(()))
+        }
+    }
+
+    // Calc most common IP and if number of most common IP exceeds need_num, return it.
+    // need_num is normally 4, 4 is half of maximum close nodes per friend.
+    // When half of clients have same IP with different port, we consider it as
+    // friend is behind NAT.
```

Comment:
You still did not fix major -> common, you did not clarify what is `most common IP`